datasets/loader_mul.py

```diff
@@ -0,0 +1,291 @@
+    mnist X train dataset (60000, 784)
+    mnist y train dataset (60000, 1)
+    mnist X test dataset  (10000,  784)
+    mnist y test dataset  (10000,  1)
+    """
+    dataset_name = 'mnist'
+
+    os.makedirs(dataset_dir, exist_ok=True)
+
+    X, y = fetch_openml(name='mnist_784', return_X_y=True,
+                        as_frame=True, data_home=dataset_dir)
+    y = y.astype(int)
+    X = X / 255
+
+    logging.info(f'{dataset_name} dataset is downloaded')
+    logging.info('reading CSV file...')
+
+    x_train, x_test, y_train, y_test = train_test_split(
+        X, y, test_size=10000, shuffle=False)
+    for data, name in zip((x_train, x_test, y_train, y_test),
+                          ('x_train', 'x_test', 'y_train', 'y_test')):
+        filename = f'{dataset_name}_{name}.csv'
+        data.to_csv(os.path.join(dataset_dir, filename),
+                    header=False, index=False)
+    logging.info(f'dataset {dataset_name} ready.')
+    return True
+
+
+def msrank(dataset_dir: Path) -> bool:
+    """
+    Dataset from szilard benchmarks: https://github.com/szilard/GBM-perf
+
+    TaskType:binclass
+    NumberOfFeatures:700
+    NumberOfInstances:10100000
+    """
+    dataset_name = 'msrank'
+    os.makedirs(dataset_dir, exist_ok=True)
+    url = "https://storage.mds.yandex.net/get-devtools-opensource/471749/msrank.tar.gz"
+    local_url = os.path.join(dataset_dir, os.path.basename(url))
+    unzipped_url = os.path.join(dataset_dir, "MSRank")
+    if not os.path.isfile(local_url):
+        logging.info(f'Started loading {dataset_name}')
+        retrieve(url, local_url)
+    if not os.path.isdir(unzipped_url):
+        logging.info(f'{dataset_name} is loaded, unzipping...')
+        tar = tarfile.open(local_url, "r:gz")
+        tar.extractall(dataset_dir)
+        tar.close()
+    logging.info(f'{dataset_name} is unzipped, started parsing...')
+
+    sets = []
+    labels = []
+    n_features = 137
+
+    for set_name in ['train.txt', 'vali.txt', 'test.txt']:
+        file_name = os.path.join(unzipped_url, set_name)
+
+        n_samples = count_lines(file_name)
+        with open(file_name, 'r') as file_obj:
+            X, y = read_libsvm_msrank(file_obj, n_samples, n_features, np.float32)
+
+        sets.append(X)
+        labels.append(y)
+
+    sets[0] = np.vstack((sets[0], sets[1]))
+    labels[0] = np.hstack((labels[0], labels[1]))
+
+    sets = [np.ascontiguousarray(sets[i]) for i in [0, 2]]
+    labels = [np.ascontiguousarray(labels[i]) for i in [0, 2]]
+
+    for data, name in zip((sets[0], sets[1], labels[0], labels[1]),
+                          ('x_train', 'x_test', 'y_train', 'y_test')):
+        filename = f'{dataset_name}_{name}.npy'
+        np.save(os.path.join(dataset_dir, filename), data)
+    logging.info(f'dataset {dataset_name} is ready.')
+    return True
+
+
+def plasticc(dataset_dir: Path) -> bool:
+    """
+    Still doesn't have an loading instruction
+    """
+    return False
```

Comment:
add `# TODO: ...` to highlight future need of dataset loader

Reply:
Done